File: Leetcode/problem167.js
// 167. Two Sum II - Input Array Is Sorted

// brute
function twoSum(numbers, target) {
    const n = numbers.length;
    for (let i = 0; i < n - 1; i++) {
        for (let j = i + 1; j < n; j++) {

            const current = numbers[i] + numbers[j];
            if (current == target)
                return [i + 1, j + 1];
        }
    }

    return [-1, -1];
}

// Test Cases
// console.log(twoSum([2, 7, 11, 15], 9));
// console.log(twoSum([2, 7, 11, 15], 26));
// console.log(twoSum([2,3,4], 6));
// console.log(twoSum([-1,0], -1));
// console.log(twoSum([1,2], 3));
// console.log(twoSum([-10,-5,-2,1,3], -7)); // [1, 5]
// console.log(twoSum([-8,-4,0,4,8], 0));
// console.log(twoSum([1,2,3,4,4,9], 8));
// console.log(twoSum([1,2,8,10,15], 3));
// console.log(twoSum([1,3,5,7,9,11], 20));
// console.log(twoSum([1,2,3,4,100], 101));

File: Leetcode_CPP/problem167.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

// brute (TLE)
// time - O(n^2)
// space - O(1)
// vector<int> twoSum(vector<int>& nums, int target) {
//     int n = nums.size();
//     for (int i = 0; i < n; i++)
//     {
//         for (int j = i + 1; j < n; j++)
//         {
//             if (nums[i] + nums[j] == target)
//             {
//                 return {i+1, j+1};
//             }
//         }
//     }

//     return {-1, -1};
// }

// better
// time - O(n) + O(log n)
// space - O(n)
// vector<int> twoSum(vector<int>& nums, int target) {
//     unordered_map<int, int> hashmap;
//     for(int i = 0; i<nums.size(); i++) {
//         int need = target-nums[i];
//         if(hashmap.count(need)) {
//             auto it = hashmap.find(need);
//             return {it->second+1, i+1};
//         }
//         hashmap[nums[i]] = i;
//     }

//     return {-1, -1};
// }

// optimal
// time - O(n)
// space - O(1)
vector<int> twoSum(vector<int>& numbers, int target) {
    int start = 0;
    int end = numbers.size() - 1;

    while (start < end) {
        int sum = numbers[start] + numbers[end];

        if(sum == target) {
            return {start+1, end+1};
        } else if(sum < target) {
            start++;
        } else end--;
    }
    
    return {-1, -1};
}

int main() {
    vector<int> arr = {2,7,11,15};

    vector<int> res = twoSum(arr, 9);
    for (auto &&it : res) {
        cout<<it<<endl;
    }
    
    return 0;
}